### ISSUE TITLE
pypiserver on AppEngine with Google Storage

### ISSUE THREAD

**Issue by deemetree-at-satelligence** [pypiserver/pypiserver#322]
# What

Branching off from https://github.com/pypiserver/pypiserver/issues/253 to coordinate this specific idea. 

## Proposal

I've forked the repository to add a basic setup allowing to host the package manager on `Google AppEngine` with `Google Storage` as the persistence: https://github.com/deemetree-at-satelligence/pypiserver/tree/deemetree-appengine-plugin  

I have it running already in own environment under a very low usage and it seems to perform reliably enough. Here I am trying to port it in a generic fashion. 

The idea in a nutshell was:  

1. `AppEngine` has only [`/tmp` storage](https://cloud.google.com/appengine/docs/standard/python3/using-temp-files) that is cleaned on each instance. 
2. `Google Storage` allows to host any files in buckets, so just as well could be used to store packages.
3. Since `pypiserver` is based on `bottle`, we can use request [`hooks`](https://bottlepy.org/docs/dev/recipes.html#using-the-hooks-plugin) to synchronize the files locally and remotely.  

I would very much appreciate the discussion and suggestions on how to move with it further. Criticism is also very welcome.

## Roadmap

More work needs to be done before it's ready. 

- [x] Document the new setup idea
- [x] Implement the basic functionality (I think that's ready, but it's been quite a while since I did it)
- [x] Make sure it's test-covered

### Working Solution

Here is a an original implementation that I created for running the pypiserver on AppEngine and Google Cloud Storage:
https://gist.github.com/deemetree-at-satelligence/793ca96bd571bba40ce1f35e9951e73b (gist)  + https://github.com/deemetree-at-satelligence/pypiserver-appengine (repo)
It's already usable to get started with your own deployments.

#### Example

![image](https://user-images.githubusercontent.com/54435277/89040348-eebf3780-d343-11ea-826d-0f423e736247.png)

**Comment by mplanchard** [comment]
Thanks for opening this ticket and helping dive in on the details! I'm at work for a bit yet today, but I'll pop back in tonight with feedback 👍

**Comment by Luttik** [comment]
@deemetree-at-satelligence you mentioned in #253 that the branch you mentioned is working fine for your team right now. For us it is also not mission critical (yet), but I'd love to give it a try already. Could you add a section to the readme that specifies what you need to configure (asside from the normal pypiserver config) to get this to run?

That would be awesome.

PS. I totally get it if you cannot find the time to explain this for lazy developers like me 😉

**Comment by mplanchard** [comment]
@deemetree-at-satelligence, I went and took a look at your branch, and I dig it. It's well organized and clean. I'm not super familiar with AppEngine (I believe I used it once at a job maybe 5 years ago), but it seems clear enough what's going on.

I particularly like how you've gotten it to sit right on top of all of the existing functionality using the bottle hooks. We've had a sort of standing TODO for quite a while to pull auth and file storage into setuptools' entrypoints, allowing easier plugin development by the community. This sidesteps that whole issue, which is great.

One thing I note is that it's _so_ well separated, that it could probably be distributed as its own package if you wanted. I'm happy to include it in the main repo of course, but it's something to keep in mind. You'd have a lot more freedom and control distributing it separately, and we could feature it in the main README. That way you'd also be isolated from my occasional "I have no time to work on side projects for three to six months at a time" spells. If you want to get it into the main repo, though, I'm also happy to add you as a contributor to make the process of supporting and expanding the plugin easier, and you'd also be able to help work on other things if you wanted.

If you want to have it included in the main repo, I agree with your checklist. I'm not sure how easy it will be to test the remote bits of it, given that we probably don't have an AppEngine instance available in the travis environment. It does support Docker, so if there's a docker container that mimics the AppEngine interface, that could work.

**Comment by deemetree-at-satelligence** [comment]
hey @mplanchard @Luttik, thanks a lot, that actually is amazing to hear! Very glad to read such a review, thanks a lot! 

wow, I was actually thinking about that because originally I developed it as an extension on top of pypiserver, so I am very glad to hear, but I think I would actually indeed prefer it to become a part of this project. For a combination of reasons - objectively at this point I wouldn't have enough time to maintain it properly on my own and certainly keeping the code for it up to date too... And more than that, thank you, in fact, I would be more than happy to be a part of an open project and become a contributor and collaborate some more, definitely, it would be a pleasure to be involved on other ideas as well.    

So yeah, if that's okay, I would be very happy for it to make it in the main repo :) Maybe later with more plugin ideas, it would be possible to detach this one and focus on providing a convenient integration interface, for example, which I would be very happy to collaborate on. But then maybe I could ask to add my main github account for this also? This is my working profile, but in my free time I usually stick to this one https://github.com/dee-me-tree-or-love

I will continue working on this next week as planned, cool. Oh, a Docker to mimic it is a nice idea, I am not yet aware if there is one, but I'll give a quick check :)  

:v: 

P.S: @Luttik, yeah, I don't have a README part yet, but you can check the example here: https://github.com/deemetree-at-satelligence/pypiserver/blob/deemetree-appengine-plugin/pypiserver/community/appengine/__init__.py#L58 it might be not so straightforward yet to tweak it to your case - I haven't adjusted everything yet - but please give it a try if you'd like.

**Comment by Luttik** [comment]
@deemetree-at-satelligence I'd also love to have a docker for this... Id also wonder if it would be possible to not only host the packages but also the .htaccessfile in a bucket. That way you would have to deal with the whole persistent volume hassle in kubernetes.

**Comment by mplanchard** [comment]
> So yeah, if that's okay, I would be very happy for it to make it in the main repo :) Maybe later with more plugin ideas, it would be possible to detach this one and focus on providing a convenient integration interface, for example, which I would be very happy to collaborate on. But then maybe I could ask to add my main github account for this also? This is my working profile, but in my free time I usually stick to this one https://github.com/dee-me-tree-or-love

Sounds good! I'll go ahead and add that account. 

> I will continue working on this next week as planned, cool.

Excellent. Feel free to post here if you run into any issues that I might be able to help with

**Comment by deemetree-at-satelligence** [comment]
@mplanchard thanks! that's great to hear, I have tried to tidy it up a little today, I will be trying to simplify it a bit more, also tried running it on my own AppEngine but that didn't work in the beginning, I will look into it later this week. 

@Luttik, I've added a few sentences in the README, so please feel free to try it out (keep in mind the code is not tested and since it diverged a bit from what is running under our own environment, there maybe be quite some bugs), but I hope that gives an idea a bit. https://github.com/deemetree-at-satelligence/pypiserver/tree/deemetree-appengine-plugin#running-in-paas

It's still all a subject to change :) for now wanted to give a small update.

**Comment by Luttik** [comment]
@deemetree-at-satelligence the added docs are super clear 👍. I'd love to find some time to test it manually (or help write tests) but I think I'll have a hard time finding spare time this week...

**Comment by Luttik** [comment]
@deemetree-at-satelligence another question. Have you changed anything about the .htaccess file? Can that also be stored in the bucket, or is that a bad idea?

**Comment by Luttik** [comment]
I tried to build a docker based on @deemetree-at-satelligence form see my [new fork here](https://github.com/Luttik/pypiserver/tree/deemetree-appengine-plugin). I have the issue that the build process seems to exclude the `community` module that @deemetree-at-satelligence added. Hence, the resulting docker will always result in a `ModuleNotFoundError: No module named 'pypiserver.community'`

I have no clue how that can be but I've spend 2.5 hours looking for a solution without any progress at all 😢. 

If I try to explicitly define `community` in `pypiserver/init.py` I get the even weirder error:
```
    ERROR: Command errored out with exit status 1:
     command: /usr/local/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-441dak3q/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-441dak3q/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open
)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-rm040xc0
         cwd: /tmp/pip-req-build-441dak3q/
    Complete output (7 lines):
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-req-build-441dak3q/setup.py", line 43, in <module>
        version=get_version(),
      File "/tmp/pip-req-build-441dak3q/setup.py", line 36, in get_version
        return d["__version__"]
    KeyError: '__version__'
    ----------------------------------------
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
```

@mplanchard I hope you can help with this issue or shine some light on how your build process works in general.

**Comment by Luttik** [comment]
For those coming here just to get PyPiServer to work with GCS buckets (to get some nice multi-read-write persistent storage) this code worked for me: [gist](https://gist.github.com/Luttik/ec8aad6ce43a26d81081c6f9df758452)

Of course, you need to add your own `.htpasswd` and google cloud service account credentials (`pypi-bucket-editor-credentials.json` the service account needs at least google cloud storage editor IIRC). 

I assume my docker isn't the best (I'm not an expert in that sense at all) so suggestions are welcome.

**Comment by deemetree-at-satelligence** [comment]
@Luttik hello! Hey, great, thanks for picking it up! 
I am still working on cleaning up the code in this fork, hence indeed it might be (unfortunately) possible that the build fails (I was getting to investigate that too last time and was exactly the moment I called it a day). Nice, nice to see you got something working with the docker, that's great :+1: 

I will be aiming at testing the setup with the App Engine deploys when I get to it using my personal GCP account. 
I will also go through your fork to see the changes :+1: 

And plus one for the investigations of the build process, I have been looking at the .travis.yml to see how does it work in the CI/CD setting and looks pretty straightforward that via this script https://github.com/pypiserver/pypiserver/blob/master/bin/gen-standalone.sh it simply runs this command `python setup.py bdist_wheel` but then modifies the wheel (which for local builds is probably not that necessary). If I am looking in the right direction, would be nice to know. @Luttik maybe that also helps somehow.

**Comment by Luttik** [comment]
@deemetree-at-satelligence are you open to a call about the current status and potential next steps? We both live in the Netherlands so it should be easy to setup. You can email me your preferred method of communication (dtluttik@gmail.com).

**Comment by deemetree-at-satelligence** [comment]
@Luttik hello! To be honest at this point I am rather busy and don't see a lot of added value in the call. I am sorry this process is going so slow but I really don't have that much free time on my side... So. Could you please explain what you are trying to achieve? If you are impatient to get it running, I have created a [gist](https://gist.github.com/deemetree-at-satelligence/793ca96bd571bba40ce1f35e9951e73b) that is based on my implementation but works _on top_ of the existing `pypiserver` (using it as dependency, which was how I originally solved the problem) using AppEngine and Google Cloud Storage. I think this would help you to get started with whatever you would like to achieve. With this issue I am indeed trying to make it a part of the package's core. I would very much appreciate your comments and feedback about the code if you have any, please of course do study it if you'd like to point something out, but at this point I don't see a big reason for a call, feel free to contact me personally if I am not understanding your point correctly. You can reach me at orlov@satelligence.com or dmitrii_orlov_devs@yahoo.com. 

Here is the gist: https://gist.github.com/deemetree-at-satelligence/793ca96bd571bba40ce1f35e9951e73b and it has the documentation how to get started with it there as well :) 

P.S: Thinking about it, I might want to make it into a separate package indeed (besides from including it here for people who'd like it separate too :shrug:). (Another note, I will be not available fully from next weekend to 25th of August for the reasons of a laptop-free vacation :) :biking_man: :mountain: )
P.P.S: apologies for a TLDR; feedback appreciated

**Comment by Luttik** [comment]
@deemetree-at-satelligence no rush. I just think that if we want to make a real effort to get this onto the main branch I think it is good to make it a coordinated effort and keep each other up to date on the architectural choices and our thought process. I don't believe github issues is the best medium for that type of communication. 

I don't have a rush since I have a working pypiserver myself (using the gist that I posted before). So I think it would be better to continue this effort when you get back from your vacation (cycling in the mountains #jealous) and when you have time. 

For the time being, I have also opened this issue https://github.com/pypiserver/pypiserver/issues/325 where I'll try to create a more general discussion on the future of the project and how we can better accommodate for contributions. 

My mindset is currently that I am willing to contribute but only if I see this project (the entire package not just your contribution) accelerating in the right direction and if there are sufficient efforts to get new contributors up to speed (i.e. I don't want my contributions going to waste on a dying package or because I cannot be working efficiently).

**Comment by dee-me-tree-or-love** [comment]
Hello everyone, back from the holidays (@deemetree-at-satelligence original-self). 

@Luttik thanks for your comments, so what I wanted to come back to before the vacation and now.  
I fully agree and actually support your enthusiasm for making this project more lively but would ask to move this discussion away from this issue as it is out of the scope of this specific proposal. 

It would be great indeed to organize a place to have talks about the future of this project and make strategic decisions and encourage more engagement, but not between the two of us at this moment with a specific meeting and roadmap plan. (not at this point at least). I would propose we can set up a Slack channel (or a comparable alternative) for community discussions, but curious what do you think @mplanchard and @Luttik. I can create one some time soon if you find it a good idea too. 

Looking at your gist, by the way, it is a very nice solution for people who rely on Docker to run pypiserver and have thought of that myself too but in case of App Engine and other PAAS solutions that does not seem feasible (please prove me wrong if necessary). You can make a pull request adding this proposal to the README or creating a separate docker file. Or you can create your own repo for it (I think [staged Docker file builds](https://docs.docker.com/develop/develop-images/multistage-build/) might be helpful), please feel free to contact me on that if you have any questions. But let's not further discuss this in this thread because that is out of scope. 

Thanks, 
Cheers, 

Dmitrii

**Comment by Luttik** [comment]
As you have seen there is now a topic for the discussion that you mentioned https://github.com/pypiserver/pypiserver/issues/325. Let's continue the off-topic conversation there.

And I think you are totally correct that the solution that I use will not work in a PAAS environment. I would not even provide this as an official solution since it is a bit of a hack (the docker needs more privileges than you usually want to provide).

**Comment by dee-me-tree-or-love** [comment]
Closing for now as it is not a priority and more things are identified for Milestone v2.0.0 :)
